File: next/plugins/instancestore/instancestore.py
"""C2: viewer.instance-store — instance state CRUD (framework §7.2 / §10.2).

Instance state = per-instance persistent state driving behavior (a chat's
roles, cwd, session ids). Free-form JSON; the schema belongs to the owning
plugin — the store never interprets values. RPC surface:
`instance:_:get/set/delete/list`. Every mutation republishes the FULL state
to the `instance-store:{plugin}:{instance}` mailbox (§5.6); delete publishes
a null value as the tombstone (mailbox has no delete primitive).
"""

from __future__ import annotations

import argparse
import asyncio
import json
import logging
import os
import signal
from pathlib import Path
from typing import Any

from sdk import Ctx, Plugin, slot

logger = logging.getLogger(__name__)

MANIFEST: dict[str, Any] = {
    "id": "instance-store",
    "version": "0.1.0",
    "slots": {"get": {}, "set": {}, "delete": {}, "list": {}},
    "emits": {"state": {}},
}


def mailbox_channel(plugin_id: str, instance: str) -> str:
    return f"instance-store:{plugin_id}:{instance}"


class InstanceStorePlugin(Plugin):
    manifest = MANIFEST

    def __init__(self, db_path: Path | None = None, **client_kwargs: Any) -> None:
        super().__init__(**client_kwargs)
        self.db_path = db_path or (Path.home() / ".view" / "instance-store.json")
        # {plugin_id: {instance: state}}
        self._data: dict[str, dict[str, Any]] = {}

    async def on_start(self) -> None:
        if self.db_path.is_file():
            self._data = json.loads(self.db_path.read_text())
            count = sum(len(instances) for instances in self._data.values())
            logger.info("loaded %d instance states from %s", count, self.db_path)

    def _save(self) -> None:
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        tmp = self.db_path.with_suffix(".tmp")
        tmp.write_text(json.dumps(self._data, ensure_ascii=False, indent=2))
        tmp.replace(self.db_path)

    @slot("instance:_:get")
    async def get(self, ctx: Ctx) -> None:
        ref = _ref(ctx)
        if ref is None:
            await ctx.respond_error("invalid_request", "missing required fields: plugin, instance")
            return
        plugin_id, instance = ref
        state = self._data.get(plugin_id, {}).get(instance)
        key = ctx.value.get("key")
        await ctx.respond(state.get(key) if (key and isinstance(state, dict)) else state)

    @slot("instance:_:set")
    async def set_(self, ctx: Ctx) -> None:
        ref = _ref(ctx)
        if ref is None:
            await ctx.respond_error("invalid_request", "missing required fields: plugin, instance")
            return
        plugin_id, instance = ref
        key = ctx.value.get("key")
        value = ctx.value.get("value")
        instances = self._data.setdefault(plugin_id, {})
        if key is None:
            if not isinstance(value, dict):
                await ctx.respond_error(
                    "invalid_request", "whole-state replace requires a JSON object value"
                )
                return
            instances[instance] = value
        else:
            state = instances.setdefault(instance, {})
            if not isinstance(state, dict):
                state = {}
                instances[instance] = state
            if value is None:
                state.pop(key, None)
            else:
                state[key] = value
        self._save()
        assert self.client is not None
        await self.client.set(mailbox_channel(plugin_id, instance), instances[instance])
        await ctx.respond({"plugin": plugin_id, "instance": instance, "state": instances[instance]})

    @slot("instance:_:delete")
    async def delete(self, ctx: Ctx) -> None:
        ref = _ref(ctx)
        if ref is None:
            await ctx.respond_error("invalid_request", "missing required fields: plugin, instance")
            return
        plugin_id, instance = ref
        existed = self._data.get(plugin_id, {}).pop(instance, None) is not None
        self._save()
        assert self.client is not None
        await self.client.set(mailbox_channel(plugin_id, instance), None)  # tombstone
        await ctx.respond({"plugin": plugin_id, "instance": instance, "existed": existed})

    @slot("instance:_:list")
    async def list_(self, ctx: Ctx) -> None:
        plugin_id = ctx.value.get("plugin") if isinstance(ctx.value, dict) else None
        if plugin_id is None:
            await ctx.respond_error("invalid_request", "missing required field: plugin")
            return
        await ctx.respond(dict(self._data.get(plugin_id, {})))

    def run(self) -> None:
        parser = argparse.ArgumentParser(description="Viewer C2 instance-store plugin")
        parser.add_argument("--kernel-ws", required=True)
        parser.add_argument("--db", default=None, help="instance-state JSON file path")
        args = parser.parse_args()
        logging.basicConfig(
            level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s: %(message)s"
        )
        if args.db is not None:
            self.db_path = Path(args.db)
        managed = os.environ.get("VIEWER_MANAGED") == "1"

        async def main() -> None:
            await self.start(args.kernel_ws, managed=managed)
            stopped = asyncio.Event()
            loop = asyncio.get_running_loop()
            for signum in (signal.SIGINT, signal.SIGTERM):
                loop.add_signal_handler(signum, stopped.set)
            await stopped.wait()
            await self.stop()

        asyncio.run(main())


def _ref(ctx: Ctx) -> tuple[str, str] | None:
    if not isinstance(ctx.value, dict):
        return None
    plugin_id, instance = ctx.value.get("plugin"), ctx.value.get("instance")
    if not plugin_id or not instance:
        return None
    return plugin_id, instance

File: internal/plugins/instancestore/instancestore.go
// Package instancestore implements per-instance state persistence.
package instancestore

import (
	"context"
	"log/slog"
	"sync"

	"viewer/internal/plugins/pluginrpc"
	"viewer/internal/plugins/storefile"
	"viewer/sdk/go/busclient"
)

const defaultFilename = "instance-store.json"

var Manifest = busclient.Manifest{
	ID: "instance-store", Version: "0.1.0",
	Slots: map[string]any{
		"get": map[string]any{}, "set": map[string]any{},
		"delete": map[string]any{}, "list": map[string]any{},
	},
	Emits: map[string]any{"state": map[string]any{}},
}

type Plugin struct {
	dbPath string
	mu     sync.Mutex
	data   map[string]map[string]any
	client *busclient.Client
}

func New(dbPath string) (*Plugin, error) {
	if dbPath == "" {
		var err error
		dbPath, err = storefile.DefaultPath(defaultFilename)
		if err != nil {
			return nil, err
		}
	}
	return &Plugin{dbPath: dbPath, data: make(map[string]map[string]any)}, nil
}

func MailboxChannel(pluginID, instance string) string {
	return "instance-store:" + pluginID + ":" + instance
}

func (p *Plugin) Start(ctx context.Context, kernelWS string, managed bool) error {
	if err := storefile.Load(p.dbPath, &p.data); err != nil {
		return err
	}
	if p.data == nil {
		p.data = make(map[string]map[string]any)
	}
	client := busclient.New(kernelWS, Manifest, busclient.WithManaged(managed))
	for pattern, handler := range map[string]func(busclient.Frame){
		"instance:_:get":    p.get,
		"instance:_:set":    p.set,
		"instance:_:delete": p.delete,
		"instance:_:list":   p.list,
	} {
		if _, err := client.Subscribe(pattern, handler); err != nil {
			_ = client.Close()
			return err
		}
	}
	p.client = client
	if err := client.Connect(ctx); err != nil {
		p.client = nil
		_ = client.Close()
		return err
	}
	slog.Info("instance-store started", "db", p.dbPath)
	return nil
}

func (p *Plugin) Close() error {
	if p.client == nil {
		return nil
	}
	return p.client.Close()
}

func (p *Plugin) get(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	value, ok := pluginrpc.Object(frame)
	pluginID, instance, valid := reference(value)
	if !ok || !valid {
		p.replyError(frame, "invalid_request", "missing required fields: plugin, instance")
		return
	}
	p.mu.Lock()
	state := p.data[pluginID][instance]
	key, hasKey := optionalString(value, "key")
	var result any = state
	if object, isObject := state.(map[string]any); hasKey && key != "" && isObject {
		result = object[key]
	}
	p.mu.Unlock()
	p.reply(frame, result)
}

func (p *Plugin) set(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	value, ok := pluginrpc.Object(frame)
	pluginID, instance, valid := reference(value)
	if !ok || !valid {
		p.replyError(frame, "invalid_request", "missing required fields: plugin, instance")
		return
	}
	key, hasKey := optionalString(value, "key")
	newValue := value["value"]
	p.mu.Lock()
	instances := p.data[pluginID]
	if instances == nil {
		instances = make(map[string]any)
		p.data[pluginID] = instances
	}
	if !hasKey {
		object, isObject := newValue.(map[string]any)
		if !isObject {
			p.mu.Unlock()
			p.replyError(frame, "invalid_request", "whole-state replace requires a JSON object value")
			return
		}
		instances[instance] = object
	} else {
		state, isObject := instances[instance].(map[string]any)
		if !isObject {
			state = make(map[string]any)
			instances[instance] = state
		}
		if newValue == nil {
			delete(state, key)
		} else {
			state[key] = newValue
		}
	}
	if err := storefile.Save(p.dbPath, p.data); err != nil {
		p.mu.Unlock()
		p.replyError(frame, "write_error", err.Error())
		return
	}
	state := cloneState(instances[instance])
	if err := p.client.Set(context.Background(), MailboxChannel(pluginID, instance), state); err != nil {
		p.mu.Unlock()
		slog.Error("publish instance mailbox", "error", err)
		return
	}
	p.mu.Unlock()
	p.reply(frame, map[string]any{"plugin": pluginID, "instance": instance, "state": state})
}

func (p *Plugin) delete(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	value, ok := pluginrpc.Object(frame)
	pluginID, instance, valid := reference(value)
	if !ok || !valid {
		p.replyError(frame, "invalid_request", "missing required fields: plugin, instance")
		return
	}
	p.mu.Lock()
	instances := p.data[pluginID]
	previous, existed := instances[instance]
	existed = existed && previous != nil
	delete(instances, instance)
	if err := storefile.Save(p.dbPath, p.data); err != nil {
		p.mu.Unlock()
		p.replyError(frame, "write_error", err.Error())
		return
	}
	if err := p.client.Set(context.Background(), MailboxChannel(pluginID, instance), nil); err != nil {
		p.mu.Unlock()
		slog.Error("publish instance tombstone", "error", err)
		return
	}
	p.mu.Unlock()
	p.reply(frame, map[string]any{"plugin": pluginID, "instance": instance, "existed": existed})
}

func (p *Plugin) list(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	value, ok := pluginrpc.Object(frame)
	pluginID, valid := stringField(value, "plugin")
	if !ok || !valid {
		p.replyError(frame, "invalid_request", "missing required field: plugin")
		return
	}
	p.mu.Lock()
	instances := make(map[string]any, len(p.data[pluginID]))
	for instance, state := range p.data[pluginID] {
		instances[instance] = cloneState(state)
	}
	p.mu.Unlock()
	p.reply(frame, instances)
}

func (p *Plugin) reply(frame busclient.Frame, result any) {
	if err := pluginrpc.Respond(p.client, frame, result); err != nil {
		slog.Error("instance-store RPC response failed", "error", err)
	}
}

func (p *Plugin) replyError(frame busclient.Frame, code, message string) {
	if err := pluginrpc.RespondError(p.client, frame, code, message); err != nil {
		slog.Error("instance-store RPC error response failed", "error", err)
	}
}

func reference(value map[string]any) (string, string, bool) {
	pluginID, pluginOK := stringField(value, "plugin")
	instance, instanceOK := stringField(value, "instance")
	return pluginID, instance, pluginOK && instanceOK && pluginID != "" && instance != ""
}

func stringField(value map[string]any, key string) (string, bool) {
	if value == nil {
		return "", false
	}
	result, ok := value[key].(string)
	return result, ok
}

func optionalString(value map[string]any, key string) (string, bool) {
	raw, exists := value[key]
	if !exists || raw == nil {
		return "", false
	}
	result, ok := raw.(string)
	return result, ok
}

func cloneState(state any) any {
	object, ok := state.(map[string]any)
	if !ok {
		return state
	}
	clone := make(map[string]any, len(object))
	for key, value := range object {
		clone[key] = value
	}
	return clone
}
